lib/client/ca_export.go

```diff
@@ -95,35 +104,99 @@ func (r *ExportAuthoritiesRequest) shouldExportIntegration(ctx context.Context)
 // For example:
 // > @cert-authority *.cluster-a ssh-rsa AAA... type=host
 // URL encoding is used to pass the CA type and allowed logins into the comment field.
-func ExportAuthorities(ctx context.Context, client authclient.ClientI, req ExportAuthoritiesRequest) (string, error) {
-	if isIntegration, err := req.shouldExportIntegration(ctx); err != nil {
-		return "", trace.Wrap(err)
-	} else if isIntegration {
-		return exportAuthForIntegration(ctx, client, req)
+//
+// At least one authority is guaranteed on success.
+func ExportAllAuthorities(ctx context.Context, client authclient.ClientI, req ExportAuthoritiesRequest) ([]*ExportedAuthority, error) {
+	const exportSecrets = false
+	return exportAllAuthorities(ctx, client, req, exportSecrets)
+}
+
+// ExportAllAuthoritiesSecrets exports authority private keys.
+// See [ExportAllAuthorities] for more information.
+//
+// At least one authority is guaranteed on success.
+func ExportAllAuthoritiesSecrets(ctx context.Context, client authclient.ClientI, req ExportAuthoritiesRequest) ([]*ExportedAuthority, error) {
+	const exportSecrets = true
+	return exportAllAuthorities(ctx, client, req, exportSecrets)
+}
+
+func exportAllAuthorities(
+	ctx context.Context,
+	client authclient.ClientI,
+	req ExportAuthoritiesRequest,
+	exportSecrets bool,
+) ([]*ExportedAuthority, error) {
+	var authorities []*ExportedAuthority
+	switch isIntegration, err := req.shouldExportIntegration(ctx); {
```

Comment:
nit: this code runs client-side and will not behave properly when we'll try to invoke it with an integration not supported by the client yet. Because `shouldExportIntegration` silently fails and swallows unknown `req.Integration`, `isIntegration` will be false and we'll try to get a unknown CA the regular way. This will likely lead lead to misleading error message and confusion. It might be better to consider an unknown `req.Integration` a hard failure in this switch.

Reply:
It looks strange to me as well, but I didn't actually change this behavior here, I only re-wrote it in an equivalent switch.

Blame points me to @greedy52 and #49396, so maybe STeve will have more context for us. I'm fine with changing this logic but I don't think the change belongs in this PR.

Reply:
sorry for the confusion. Integration CAs are saved as plugin credentials instead of the regular CAs. They don't support CA rotation, multiple active keys, etc. 

I can refactor integration CAs outside this function after this change. What do you think?

Reply:
Could we hard-fail in this branch: https://github.com/gravitational/teleport/blob/f55257a110df8bdb7d678bfb367c96478a212992/lib/client/ca_export.go#L65-L71

If yes I think that addresses the comment. No strong opinions on refactors otherwise.

Reply:
yes